### PR TITLE
Fix #773: Implementation of mloginfo --sharding 

### LINKED ISSUE #773: mloginfo --sharding
- Support parsing mongod/mongos logs from version 3.6 to 4.2 for sharding related information.
- Output sharding related configurations:
- Output errors/warnings related to sharding
- Output chunk migration statistics for each namespace
- Output chunk split statistics

### PR DESCRIPTION
**Pull request by Giuliano-C**
## Description of Improvement
Added --sharding extension to mloginfo, which outputs:

- Sharding configuration overview
- Sharding related errors/warnings
- Chunk migrations (shard being destination and/or source)
- Chunk split statistics

Tests have been added to ensure correct information is being outputted. Documentation has also been updated to outline how the --sharding flag works

### CODE REVIEW

File: mtools/test/logfiles/sharding_379_mongos.log [hunk 1]
```diff
@@ -0,0 +1,78 @@
+2020-01-28T14:26:53.776+1100 I CONTROL  [main] 
+2020-01-28T14:26:53.777+1100 I CONTROL  [main] ** NOTE: This is a development version (3.7.9) of MongoDB.
```
Comment (stennie) [inline mtools/test/logfiles/sharding_379_mongos.log]: @Giuliano-C Why is this test generated using an older development version of MongoDB?

We are only supporting stable/production releases (currently MongoDB 3.6, 4.0, and 4.2). If a test requires changes committed during the 3.7 development cycle, the test case should be the next production release that includes the change (so 4.0.0).

File: doc/index.rst [hunk 2]
```diff
@@ -74,4 +74,4 @@ posted in the `Issues
 .. |Python 37| image:: https://img.shields.io/badge/Python-3.7-brightgreen.svg?style=flat
    :target: http://python.org
 .. |Python 38| image:: https://img.shields.io/badge/Python-3.8-brightgreen.svg?style=flat
-   :target: http://python.org
\ No newline at end of file
+   :target: http://python.org
```
Comment (stennie) [inline doc/index.rst]: Was this (and similar) doc changes for line length? It is generally better to keep cosmetic/unrelated changes separate for clarity of a feature patch.

File: doc/mloginfo.rst [hunk 3]
```diff
@@ -393,11 +396,106 @@ For example:
...
 
+Sharding (``--sharding``)
+-----------------------------------------
+
+Outputs information of sharding configuration. Includes groupings of
+similar error/warning messages that are specific to sharding. A breakdown
+of chunk migration activity is included, specifying the source/destination
+shard, the number of chunk migrations attempted within the hour, and the
+ratio of successful to failed migrations. Failed chunk migrations specify
+when the failure occurred in formatted as [hours:minutes:seconds:milliseconds]
```
Comment (stennie) [inline doc/mloginfo.rst]: The "Failed chunk migrations ..." sentence needs some rephrasing for clarity. Since you are going to show an example below you could trim some of the details like {hours:minutes:..) formatting.

I'm also unsure what "precedes to be successful" means.

File: doc/mloginfo.rst [hunk 4]
```diff
@@ -393,11 +396,106 @@ For example:
...
+time frame). It also outputs chunk split statistics, which has a similar output
+as the chunk migration tables.
+
+For example:
+
+.. code-block:: bash
+
+   mloginfo mongod.log --sharding
+
+In addition to the default information, this command also outputs sharding
```
Comment (stennie) [inline doc/mloginfo.rst]: This appears to repeat most of the description above (starting from line 402). I would make this more concise and just show the example output.

You could also move some of the detail above (like description of the timestamp format) as further annotation for the example output.

File: doc/mloginfo.rst [hunk 5]
```diff
@@ -393,11 +396,106 @@ For example:
...
+         shard01: example:27018,example:27019,example:27020
+         shard02: example:27021,example:27022,example:27023
+         shard03: example:27024,example:27025,example:27026
+      CSRS:
+         configRepl: example:27033
+
+``--errors``
+^^^^^^^^^^^^
+
+This option toggles if sharding related errors/warnings are
```
Comment (stennie) [inline doc/mloginfo.rst]: "This option toggles" sounds like the option will be triggered automatically. It would be clearly to describe the action, eg: "This option can be added to `--sharding` to include a summary of any sharding-related errors or warnings".

Since errors are normally hard fails and warnings are not ... this may want to be renamed or split into several options such as `--errors`, `--warnings`, and `--sharding-verbose` (which could imply all sharding details).

File: doc/mloginfo.rst [hunk 6]
```diff
@@ -393,11 +396,106 @@ For example:
...
+outputted.
```
Comment (stennie) [inline doc/mloginfo.rst]: This should be present tense "output"

File: doc/mloginfo.rst [hunk 7]
```diff
@@ -393,11 +396,106 @@ For example:
...
+   Error Messages:
+
+      22  ... Deletion of XXX range [...) will be scheduled after all possibly dependent queries finish
+
+This option has no effect unless ``--sharding`` is also specified.
+
+``--migrations``
+^^^^^^^^^^^^^^^^
+
+This option toggles if chunk migrations and split statistics are
```
Comment (stennie) [inline doc/mloginfo.rst]: Similar to above comment on `--errors`, would describe without using toggles.

File: mtools/test/test_mloginfo.py [hunk 8]
```diff
@@ -400,6 +400,109 @@ def _test_rsinfo(self, logfile_path, **expected):
             print("results[%s] == %s" % (key, value))
             assert results.get(key) == value
 
+    def test_sharding_output(self):
+        self.tool.run('--sharding --errors --migrations %s' % self.logfile_path)
+        output = sys.stdout.getvalue()
+        lines = output.splitlines()
+        assert any(map(lambda line: 'SHARDING' in line, lines))
+
+        assert any(map(lambda line: 'Overview:' in line, lines))
```
Comment (stennie) [inline mtools/test/test_mloginfo.py]: Test should generally be validating that the expected analysis has been performed. You could use the section headings as an anchor to find the expected details, but the important part of the test would be making sure that data like error messages and number of chunks moved from/to matches expectations.

File: mtools/test/test_mloginfo.py [hunk 9]
```diff
@@ -400,6 +400,109 @@ def _test_rsinfo(self, logfile_path, **expected):
...
+                                    'localhost:27020' in line, lines))
+        assert any(map(lambda line: 'configRepl' in line, lines))
+        assert any(map(lambda line: 'localhost:27033' in line, lines))
+
+    def test_sharding_error_messages_exists(self):
+        self._test_init('sharding_379_shard.log')
+        self.tool.run('--sharding --errors %s' % self.logfile_path)
+        output = sys.stdout.getvalue()
+        lines = output.splitlines()
+        assert len(list(filter(lambda line: re.match('^  \d .* ', line),
```
Comment (stennie) [inline mtools/test/test_mloginfo.py]: This test needs a comment explaining what the regex is trying to match (eg example of matching output line)

File: mtools/test/test_mloginfo.py [hunk 10]
```diff
@@ -400,6 +400,109 @@ def _test_rsinfo(self, logfile_path, **expected):
...
+                          lines))) == 2
+
+    def test_sharding_chunk_migration_from_exists(self):
+        self._test_init('sharding_379_shard.log')
+        self._test_sharding(self.logfile_path, r'  2020-01-28T16', 3)
```
Comment (stennie) [inline mtools/test/test_mloginfo.py]: The `_test_sharding()` parameters would be less mysterious if you change to using named parameters (`pattern`, `expected`).

File: mtools/test/test_util_logfile.py [hunk 11]
```diff
@@ -114,3 +114,90 @@ def test_hostname_port(self):
         print(logfile2.hostname)
         assert logfile2.hostname == 'jimoleary.local'
         assert logfile2.port == '27017'
+
+    def test_shard_info(self):
+        """LogFile: test if sharding info is detected (MongoDB 3.4+) """
```
Comment (stennie) [inline mtools/test/test_util_logfile.py]: If this should work in 3.4+, I'd use 3.4 as the baseline (or 3.6 since 3.4 is EOL)

File: mtools/util/logfile.py [hunk 12]
```diff
@@ -206,6 +214,41 @@ def storage_engine(self):
             self._iterate_lines()
         return self._storage_engine
 
+    @property
+    def shards(self):
+        """Return the shards (if available)"""
+        if not self._num_lines:
+            self._iterate_lines()
```
Comment (stennie) [inline mtools/util/logfile.py]: The sharding details should ideally be in a separate lazy method rather than `_iterate_lines()`, so extra matches are only run when needed.

File: doc/index.rst (hunk 2 above)
Comment (Giuliano-C) [reply doc/index.rst]: I removed cosmetic/unrelated changes from history

File: doc/mloginfo.rst (hunk 3 above)
Comment (Giuliano-C) [reply doc/mloginfo.rst]: I removed the sentences that wasn't necessary

File: doc/mloginfo.rst (hunk 4 above)
Comment (Giuliano-C) [reply doc/mloginfo.rst]: I made the outline more succinct

File: doc/mloginfo.rst (hunk 5 above)
Comment (Giuliano-C) [reply doc/mloginfo.rst]: I re-worded to indicate its an addition of information when added with `--sharding`

File: mtools/test/test_mloginfo.py (hunk 8 above)
Comment (Giuliano-C) [reply mtools/test/test_mloginfo.py]: I decided to remove this test and add more detail to others so that the table expectations were tested more

File: mtools/util/logfile.py (hunk 12 above)
Comment (Giuliano-C) [reply mtools/util/logfile.py]: Created a seperate lazy method to find all the sharding information
